### PR TITLE
feat: noImplicitAny

### PR DESCRIPTION
**Pull request by pedromtec**
## What's the purpose of this pull request?

Strict mode in TypeScript is a way to opt into a more strict type-checking behavior to catch more potential issues at compile time. When enabled, it activates a set of stricter type-checking options that enforce more constraints on your code. Here's how it works and what it does:

**No Implicit Any (noImplicitAny)**: This prevents variables and parameters from defaulting to the any type when TypeScript is unable to infer a more specific type. This helps catch cases where you might have forgotten to specify a type or TypeScript can't infer one for you.

To enable strict mode, you can set "**strict": true** in your `tsconfig.json` file. This is equivalent to enabling all of the above options individually.

The idea of ​​this initiative is to enable rule by rule incrementally so that it is possible to open smaller PRs with low risk.

Task: https://vtex-dev.atlassian.net/browse/SFS-1124

## How it works?

The first PR enables `noImplicitAny` rule. So codes like that will not work: 

What's the type of address? 
```
function printAddresss(address) {
    console.log(address) 
}
``` 

## How to test it?

- The pipeline (build, tests, and lint) should work for **@faststore** packages and their dependencies (current branch). 

- Install the version of this PR in the [starter.store](https://github.com/vtex-sites/starter.store) and test if the build, tests and override feature works. 

"@faststore/core": "https://pkg.csb.dev/vtex/faststore/commit/1b7aa312/@faststore/core"
"@faststore/cli": "https://pkg.csb.dev/vtex/faststore/commit/1b7aa312/@faststore/cli",

I welcome other suggestions and help with testing 🥇

### CODE REVIEW

File: packages/core/src/pages/[slug]/p.tsx [hunk 1]
```diff
@@ -63,7 +63,7 @@ type Props = PDPContentType & {
 
 // Array merging strategy from deepmerge that makes client arrays overwrite server array
 // https://www.npmjs.com/package/deepmerge
-const overwriteMerge = (_, sourceArray) => sourceArray
+const overwriteMerge = (_: any[], sourceArray: any[]) => sourceArray
```
Comment (pedromtec) [inline packages/core/src/pages/[slug]/p.tsx:66]: `any[]` is the type that `deepmerge` library expects

File: packages/core/src/components/ui/Newsletter/NewsletterAddendum.tsx [hunk 2]
```diff
@@ -13,6 +13,8 @@
  *
  * This is a limitation not only for this component, but for every native & custom component that makes use of Rich Text.
  */
+
+// @ts-ignore motivation: cannot find the draftjs-to-html package types.
```
Comment (pedromtec) [inline packages/core/src/components/ui/Newsletter/NewsletterAddendum.tsx:17]: The comment above says that this library will be removed soon. Her types are missing. I could try to install or create a d.ts file. However, the idea is that the library is removed

File: packages/core/src/components/ui/Image/loader.ts [hunk 3]
```diff
@@ -1,7 +1,15 @@
...
 
-export default function customImageLoader({ src, width, quality }) {
+export default function customImageLoader({
+  src,
+  width,
+  quality,
+}: {
+  src: any
+  width: any
+  quality: any
```
[the remark this answers is not in the document]
Comment (pedromtec) [reply packages/core/src/components/ui/Image/loader.ts]: Fixed

File: packages/cli/src/utils/generate.ts [hunk 4]
```diff
@@ -60,6 +60,31 @@ function filterAndCopyPackageJson(basePath: string) {
...
 
+// Temporary array of strict rules enabled so far.
+const TS_CONFIG_STRICT_RULES_ENABLED = ['noImplicitAny'] as const
+
+/**
+ * Modify TypeScript compilation settings (tsconfig.json) to disable specific strict
+ * type checking rules when files are moved to the .faststore folder. 
+ * When all strict rules are migrated. The idea is to only change the strict to false
+ */
+function disableTsConfigStrictRules(basePath: string) {
```
Comment (pedromtec) [inline packages/cli/src/utils/generate.ts:71]: To avoid problems when building stores, strict mode is disabled. It is only needed for people who maintain the `@faststore/core` project

File: packages/cli/src/utils/generate.ts [hunk 5]
```diff
@@ -60,6 +60,31 @@ function filterAndCopyPackageJson(basePath: string) {
   })
 }
 
+// Temporary array of strict rules enabled so far.
+const TS_CONFIG_STRICT_RULES_ENABLED = ['noImplicitAny'] as const
+
+/**
+ * Modify TypeScript compilation settings (tsconfig.json) to disable specific strict
+ * type checking rules when files are moved to the .faststore folder. 
+ * When all strict rules are migrated. The idea is to only change the strict to false
```
Comment (lariciamota) [inline packages/cli/src/utils/generate.ts:69]: suggestion:
```suggestion
 * The idea is to change the strict to false when all strict rules are migrated.
```

File: packages/core/src/components/search/SearchDropdown/SearchDropdown.tsx [hunk 6]
```diff
@@ -32,14 +37,14 @@ function SearchDropdown({ sort, ...otherProps }) {
             linkProps={{
               href: formatSearchPath({
                 term: suggestion,
-                sort: sort as SearchState['sort'],
+                sort: sort,
```
[the remark this answers is not in the document]
Comment (pedromtec) [reply packages/core/src/components/search/SearchDropdown/SearchDropdown.tsx:40]: > suggestion:

💯

File: packages/core/src/components/search/SearchInput/SearchInput.tsx [hunk 7]
```diff
@@ -133,7 +134,7 @@ const SearchInput = forwardRef<SearchInputRef, SearchInputProps>(
 
         {searchDropdownVisible && (
           <Suspense fallback={null}>
-            <SearchDropdown sort={sort} />
+            <SearchDropdown sort={sort as SearchState['sort']} />
```
[the remark this answers is not in the document]
Comment (pedromtec) [reply packages/core/src/components/search/SearchInput/SearchInput.tsx:137]: We need to define the types of all functions and components. For this reason, I had to define the props of the `SearchDropdown` component like this:

```
interface SearchDropdownProps {
  sort: SearchState['sort']
  [key: string]: any
}
```

[the remark this answers is not in the document]
Comment (pedromtec) [reply packages/core/src/components/search/SearchInput/SearchInput.tsx:137]: So I just changed the order of the `cast`.

File: packages/core/src/sdk/overrides/overrides.ts [hunk 8]
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck
```
[the remark this answers is not in the document]
Comment (pedromtec) [reply packages/core/src/sdk/overrides/overrides.ts:1]: Good question, There are 3 errors in override typing in this file. It's very complicated to try to fix things without the full context of the feature. **ts-nocheck** will disable the checking of the entire file, but thinking about it this is not a good alternative as this will omit other typescript checks.

[the remark this answers is not in the document]
Comment (pedromtec) [reply packages/core/src/sdk/overrides/overrides.ts:1]: It is quite complicated to enable these rules in an ongoing project because we can find files with problems that require some context to resolve. But most of the files were easy to resolve and I believe that this rule will improve our development and the quality of our project. I'll think of a better alternative for this file.

[the remark this answers is not in the document]
Comment (pedromtec) [reply packages/core/src/sdk/overrides/overrides.ts:1]: I found an alternative to this issue: `@ts-expect-error`

Some definitions:
**Purpose**: To explicitly indicate that you expect a TypeScript error on the following line, and to have TypeScript ensure that an error is present. If no error is found on that line, TypeScript will raise a warning.

**@ts-expect-error** is useful for scenarios where you are working around a known issue, such as a temporary type mismatch that you plan to address later. It helps maintain awareness that the line is supposed to produce an error.

**@ts-ignore** can sometimes be used inappropriately, suppressing errors that you might actually want to address.

[the remark this answers is not in the document]
Comment (pedromtec) [reply packages/core/src/sdk/overrides/overrides.ts:1]: I think we can use **@ts-expect-error**  before the lines of codes with error and complement it with some explanation of the motivation. This will only affect the line with error. WDTY?

[the remark this answers is not in the document]
Comment (pedromtec) [reply packages/core/src/sdk/overrides/overrides.ts:1]: I think we can use this approach  here too: https://github.com/vtex/faststore/pull/2379#discussion_r1701780247

File: packages/cli/src/utils/generate.ts [hunk 9]
```diff
@@ -60,6 +60,31 @@ function filterAndCopyPackageJson(basePath: string) {
   })
 }
 
+// Temporary array of strict rules enabled so far.
+const TS_CONFIG_STRICT_RULES_ENABLED = ['noImplicitAny'] as const
+
+/**
+ * Modify TypeScript compilation settings (tsconfig.json) to disable specific strict
+ * type checking rules when files are moved to the .faststore folder. 
+ * The idea is to change the strict to false when all strict rules are migrated.
```
Comment (hellofanny) [inline packages/cli/src/utils/generate.ts:69]: If I got it right, we will remove this part when all rules are migrated, right?
**suggestion**: maybe we can add a todo markup here - // TODO:

File: packages/core/src/sdk/overrides/overrides.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (lariciamota) [reply packages/core/src/sdk/overrides/overrides.ts:1]: Yes, I agree

**Review by hellofanny — APPROVED** [review]
LGTM 🚀

**Comment by pedromtec** [comment]
Team (small change). I changed from **ts-expect-error** to **ts-ignore** because **ts-expect-error** generates an error when the line has no error. In other words, it always expects the line to have an error. However, when I disable **noImplictyAny**, the lines stop having errors within the **.faststore**.

File: packages/core/src/components/search/SearchDropdown/SearchDropdown.tsx [hunk 10]
```diff
@@ -14,7 +14,12 @@ import { ProductSummary_ProductFragment } from '@generated/graphql'
 import SearchProductItem from 'src/components/search/SearchProductItem'
 import { formatSearchPath } from 'src/sdk/search/formatSearchPath'
 
-function SearchDropdown({ sort, ...otherProps }) {
+interface SearchDropdownProps {
+  sort: SearchState['sort']
+  [key: string]: any
```
Comment (lucasfp13) [inline packages/core/src/components/search/SearchDropdown/SearchDropdown.tsx:19]: **question: (non-blocking)** Why is this necessary? Since we are spreading `otherProps` I couldn't understand why this is needed

**Review by lucasfp13 — APPROVED** [review]

File: packages/core/src/components/search/SearchDropdown/SearchDropdown.tsx (hunk 10 above)
Comment (pedromtec) [reply packages/core/src/components/search/SearchDropdown/SearchDropdown.tsx:19]: The **Typescript** needs to know the typing of all props. It was difficult for me to know what the props were because of the spread (...otherProps). So the "workaround" I had to do is that the other props can be any pair with the key string and the value any
